rustdoc ice: 'internal error: entered unreachable code', compiler/rustc_resolve/src/lib.rs:3305:6

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

This code needs to be placed into `compiler/rustc_borrowck/src/diagnostics/mod.rs` (just overwrite the file).
My repo is currently at f7b4824731483df9ac911cb5b60fe85f04162ad0

```rust
//! hello
mod find_all_local_uses;
```

If you then run `rustdoc ./compiler/rustc_borrowck/src/diagnostics/mod.rs`, rustdoc will crash:

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustdoc 1.62.0-nightly (8bf93e9b6 2022-04-09)
binary: rustdoc
commit-hash: 8bf93e9b6791acee3a594ed202fdfe45ad9952b5
commit-date: 2022-04-09
host: x86_64-unknown-linux-gnu
release: 1.62.0-nightly
LLVM version: 14.0.0
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary>Backtrace</summary>
<p>

```
error[E0433]: failed to resolve: maybe a missing crate `rustc_middle`?
 --> ./compiler/rustc_borrowck/src/diagnostics/find_all_local_uses.rs:3:5
  |
3 | use rustc_middle::mir::visit::{PlaceContext, Visitor};
  |     ^^^^^^^^^^^^ maybe a missing crate `rustc_middle`?

error[E0433]: failed to resolve: maybe a missing crate `rustc_middle`?
 --> ./compiler/rustc_borrowck/src/diagnostics/find_all_local_uses.rs:4:5
  |
4 | use rustc_middle::mir::{Body, Local, Location};
  |     ^^^^^^^^^^^^ maybe a missing crate `rustc_middle`?

error[E0412]: cannot find type `Body` in this scope
 --> ./compiler/rustc_borrowck/src/diagnostics/find_all_local_uses.rs:9:33
  |
9 | pub(super) fn find<'tcx>(body: &Body<'tcx>, local: Local) -> BTreeSet<Location> {
  |                                 ^^^^ not found in this scope

error[E0412]: cannot find type `Local` in this scope
 --> ./compiler/rustc_borrowck/src/diagnostics/find_all_local_uses.rs:9:52
  |
9 | pub(super) fn find<'tcx>(body: &Body<'tcx>, local: Local) -> BTreeSet<Location> {
  |                                                    ^^^^^ not found in this scope

error[E0412]: cannot find type `Location` in this scope
 --> ./compiler/rustc_borrowck/src/diagnostics/find_all_local_uses.rs:9:71
  |
9 | pub(super) fn find<'tcx>(body: &Body<'tcx>, local: Local) -> BTreeSet<Location> {
  |                                                                       ^^^^^^^^ not found in this scope
  |
help: consider importing this struct
  |
1 | use std::panic::Location;
  |

error[E0412]: cannot find type `Local` in this scope
  --> ./compiler/rustc_borrowck/src/diagnostics/find_all_local_uses.rs:16:16
   |
16 |     for_local: Local,
   |                ^^^^^ not found in this scope

error[E0412]: cannot find type `Location` in this scope
  --> ./compiler/rustc_borrowck/src/diagnostics/find_all_local_uses.rs:17:20
   |
17 |     uses: BTreeSet<Location>,
   |                    ^^^^^^^^ not found in this scope
   |
help: consider importing this struct
   |
1  | use std::panic::Location;
   |

error[E0405]: cannot find trait `Visitor` in this scope
  --> ./compiler/rustc_borrowck/src/diagnostics/find_all_local_uses.rs:20:12
   |
20 | impl<'tcx> Visitor<'tcx> for AllLocalUsesVisitor {
   |            ^^^^^^^ not found in this scope

error[E0412]: cannot find type `Local` in this scope
  --> ./compiler/rustc_borrowck/src/diagnostics/find_all_local_uses.rs:21:39
   |
21 |     fn visit_local(&mut self, local: &Local, _context: PlaceContext, location: Location) {
   |                                       ^^^^^ not found in this scope

error[E0412]: cannot find type `PlaceContext` in this scope
  --> ./compiler/rustc_borrowck/src/diagnostics/find_all_local_uses.rs:21:56
   |
21 |     fn visit_local(&mut self, local: &Local, _context: PlaceContext, location: Location) {
   |                                                        ^^^^^^^^^^^^ not found in this scope

error[E0412]: cannot find type `Location` in this scope
  --> ./compiler/rustc_borrowck/src/diagnostics/find_all_local_uses.rs:21:80
   |
21 |     fn visit_local(&mut self, local: &Local, _context: PlaceContext, location: Location) {
   |                                                                                ^^^^^^^^ not found in this scope
   |
help: consider importing this struct
   |
1  | use std::panic::Location;
   |

thread 'rustc' panicked at 'internal error: entered unreachable code', compiler/rustc_resolve/src/lib.rs:3305:67
stack backtrace:
   0:     0x7ff4d549e00d - std::backtrace_rs::backtrace::libunwind::trace::h06b71e168fca8ca7
                               at /rustc/8bf93e9b6791acee3a594ed202fdfe45ad9952b5/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7ff4d549e00d - std::backtrace_rs::backtrace::trace_unsynchronized::hd5c568c45ebf115f
                               at /rustc/8bf93e9b6791acee3a594ed202fdfe45ad9952b5/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7ff4d549e00d - std::sys_common::backtrace::_print_fmt::h2207f18151caed70
                               at /rustc/8bf93e9b6791acee3a594ed202fdfe45ad9952b5/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x7ff4d549e00d - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hec1ef25b87918dac
                               at /rustc/8bf93e9b6791acee3a594ed202fdfe45ad9952b5/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x7ff4d54f7d2c - core::fmt::write::h3bdfdbc9caaeb5c4
                               at /rustc/8bf93e9b6791acee3a594ed202fdfe45ad9952b5/library/core/src/fmt/mod.rs:1194:17
   5:     0x7ff4d548f8d1 - std::io::Write::write_fmt::h75f007d851ff8f3e
                               at /rustc/8bf93e9b6791acee3a594ed202fdfe45ad9952b5/library/std/src/io/mod.rs:1655:15
   6:     0x7ff4d54a0f25 - std::sys_common::backtrace::_print::h8cb608922c1cc56d
                               at /rustc/8bf93e9b6791acee3a594ed202fdfe45ad9952b5/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x7ff4d54a0f25 - std::sys_common::backtrace::print::h69d750eeeb320760
                               at /rustc/8bf93e9b6791acee3a594ed202fdfe45ad9952b5/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x7ff4d54a0f25 - std::panicking::default_hook::{{closure}}::h220d8c7b31c1776e
                               at /rustc/8bf93e9b6791acee3a594ed202fdfe45ad9952b5/library/std/src/panicking.rs:295:22
   9:     0x7ff4d54a0b99 - std::panicking::default_hook::hd732eb9b63a7d08e
                               at /rustc/8bf93e9b6791acee3a594ed202fdfe45ad9952b5/library/std/src/panicking.rs:314:9
  10:     0x7ff4d5ccf3c1 - rustc_driver[3655e38759671b1e]::DEFAULT_HOOK::{closure#0}::{closure#0}
  11:     0x7ff4d54a16c0 - std::panicking::rust_panic_with_hook::he731ea0e4d942ce4
                               at /rustc/8bf93e9b6791acee3a594ed202fdfe45ad9952b5/library/std/src/panicking.rs:702:17
  12:     0x7ff4d54a14b9 - std::panicking::begin_panic_handler::{{closure}}::h7572f87a59230791
                               at /rustc/8bf93e9b6791acee3a594ed202fdfe45ad9952b5/library/std/src/panicking.rs:586:13
  13:     0x7ff4d549e4c4 - std::sys_common::backtrace::__rust_end_short_backtrace::ha3eefb66086a857c
                               at /rustc/8bf93e9b6791acee3a594ed202fdfe45ad9952b5/library/std/src/sys_common/backtrace.rs:138:18
  14:     0x7ff4d54a1229 - rust_begin_unwind
                               at /rustc/8bf93e9b6791acee3a594ed202fdfe45ad9952b5/library/std/src/panicking.rs:584:5
  15:     0x7ff4d5465773 - core::panicking::panic_fmt::h4508dc135e264600
                               at /rustc/8bf93e9b6791acee3a594ed202fdfe45ad9952b5/library/core/src/panicking.rs:142:14
  16:     0x7ff4d546563d - core::panicking::panic::hed1ca88aad7c3049
                               at /rustc/8bf93e9b6791acee3a594ed202fdfe45ad9952b5/library/core/src/panicking.rs:48:5
  17:     0x7ff4d6130459 - <rustc_resolve[577bf44bd1426e45]::Resolver>::resolve_rustdoc_path
  18:     0x558d3f747622 - <rustdoc[31e9a93f42be1f5]::passes::collect_intra_doc_links::early::EarlyDocLinkResolver>::load_links_in_attrs
  19:     0x558d3f747ce0 - <rustdoc[31e9a93f42be1f5]::passes::collect_intra_doc_links::early::EarlyDocLinkResolver as rustc_ast[139425a543286658]::visit::Visitor>::visit_item
  20:     0x558d3f79bd2b - rustc_ast[139425a543286658]::visit::walk_item::<rustdoc[31e9a93f42be1f5]::passes::collect_intra_doc_links::early::EarlyDocLinkResolver>
  21:     0x558d3f747d2e - <rustdoc[31e9a93f42be1f5]::passes::collect_intra_doc_links::early::EarlyDocLinkResolver as rustc_ast[139425a543286658]::visit::Visitor>::visit_item
  22:     0x558d3f74559a - rustdoc[31e9a93f42be1f5]::passes::collect_intra_doc_links::early::early_resolve_intra_doc_links
  23:     0x558d3f63962d - <rustc_interface[aab2741c181344a3]::interface::Compiler>::enter::<rustdoc[31e9a93f42be1f5]::main_options::{closure#0}::{closure#0}, core[491626612f8df212]::result::Result<(), rustc_errors[ced839e2d504a56c]::ErrorGuaranteed>>
  24:     0x558d3f7d44f6 - rustc_span[2556154be3375056]::with_source_map::<core[491626612f8df212]::result::Result<(), rustc_errors[ced839e2d504a56c]::ErrorGuaranteed>, rustc_interface[aab2741c181344a3]::interface::create_compiler_and_run<core[491626612f8df212]::result::Result<(), rustc_errors[ced839e2d504a56c]::ErrorGuaranteed>, rustdoc[31e9a93f42be1f5]::main_options::{closure#0}>::{closure#1}>
  25:     0x558d3f670a04 - rustc_interface[aab2741c181344a3]::interface::create_compiler_and_run::<core[491626612f8df212]::result::Result<(), rustc_errors[ced839e2d504a56c]::ErrorGuaranteed>, rustdoc[31e9a93f42be1f5]::main_options::{closure#0}>
  26:     0x558d3f4cd237 - <scoped_tls[8f93d9f95e89128f]::ScopedKey<rustc_span[2556154be3375056]::SessionGlobals>>::set::<rustdoc[31e9a93f42be1f5]::main_args::{closure#0}, core[491626612f8df212]::result::Result<(), rustc_errors[ced839e2d504a56c]::ErrorGuaranteed>>
  27:     0x558d3f6712df - std[4f0b898746aae38d]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[aab2741c181344a3]::util::run_in_thread_pool_with_globals<rustdoc[31e9a93f42be1f5]::main_args::{closure#0}, core[491626612f8df212]::result::Result<(), rustc_errors[ced839e2d504a56c]::ErrorGuaranteed>>::{closure#0}, core[491626612f8df212]::result::Result<(), rustc_errors[ced839e2d504a56c]::ErrorGuaranteed>>
  28:     0x558d3f713a19 - <<std[4f0b898746aae38d]::thread::Builder>::spawn_unchecked_<rustc_interface[aab2741c181344a3]::util::run_in_thread_pool_with_globals<rustdoc[31e9a93f42be1f5]::main_args::{closure#0}, core[491626612f8df212]::result::Result<(), rustc_errors[ced839e2d504a56c]::ErrorGuaranteed>>::{closure#0}, core[491626612f8df212]::result::Result<(), rustc_errors[ced839e2d504a56c]::ErrorGuaranteed>>::{closure#1} as core[491626612f8df212]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  29:     0x7ff4d54ab823 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h6aa3defc4f10dc70
                               at /rustc/8bf93e9b6791acee3a594ed202fdfe45ad9952b5/library/alloc/src/boxed.rs:1858:9
  30:     0x7ff4d54ab823 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h553d7c8698383f07
                               at /rustc/8bf93e9b6791acee3a594ed202fdfe45ad9952b5/library/alloc/src/boxed.rs:1858:9
  31:     0x7ff4d54ab823 - std::sys::unix::thread::Thread::new::thread_start::h0643c814f607f90a
                               at /rustc/8bf93e9b6791acee3a594ed202fdfe45ad9952b5/library/std/src/sys/unix/thread.rs:108:17
  32:     0x7ff4d519b5c2 - start_thread
  33:     0x7ff4d5220584 - __clone
  34:                0x0 - <unknown>

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.62.0-nightly (8bf93e9b6 2022-04-09) running on x86_64-unknown-linux-gnu

query stack during panic:
end of query stack
error: aborting due to 11 previous errors

Some errors have detailed explanations: E0405, E0412, E0433.
For more information about an error, try `rustc --explain E0405`.
```

</p>
</details>

--- a/compiler/rustc_resolve/src/imports.rs
+++ b/compiler/rustc_resolve/src/imports.rs
@@ -583,18 +583,23 @@ impl<'a> Resolver<'a> {
         t
     }
 
-    // Define a "dummy" resolution containing a Res::Err as a placeholder for a
-    // failed resolution
+    // Define a dummy resolution containing a `Res::Err` as a placeholder for a failed resolution,
+    // also mark such failed imports as used to avoid duplicate diagnostics.
     fn import_dummy_binding(&mut self, import: &'a Import<'a>) {
-        if let ImportKind::Single { target, .. } = import.kind {
+        if let ImportKind::Single { target, ref target_bindings, .. } = import.kind {
+            if target_bindings.iter().any(|binding| binding.get().is_some()) {
+                return; // Has resolution, do not create the dummy binding
+            }
             let dummy_binding = self.dummy_binding;
             let dummy_binding = self.import(dummy_binding, import);
             self.per_ns(|this, ns| {
                 let key = this.new_key(target, ns);
                 let _ = this.try_define(import.parent_scope.module, key, dummy_binding);
             });
-            // Consider erroneous imports used to avoid duplicate diagnostics.
             self.record_use(target, dummy_binding, false);
+        } else if import.imported_module.get().is_none() {
+            import.used.set(true);
+            self.used_imports.insert(import.id);
         }
     }
 }
@@ -659,7 +664,13 @@ impl<'a, 'b> ImportResolver<'a, 'b> {
             .map(|i| (false, i))
             .chain(indeterminate_imports.into_iter().map(|i| (true, i)))
         {
-            if let Some(err) = self.finalize_import(import) {
+            let unresolved_import_error = self.finalize_import(import);
+
+            // If this import is unresolved then create a dummy import
+            // resolution for it so that later resolve stages won't complain.
+            self.r.import_dummy_binding(import);
+
+            if let Some(err) = unresolved_import_error {
                 if let ImportKind::Single { source, ref source_bindings, .. } = import.kind {
                     if source.name == kw::SelfLower {
                         // Silence `unresolved import` error if E0429 is already emitted
@@ -669,9 +680,6 @@ impl<'a, 'b> ImportResolver<'a, 'b> {
                     }
                 }
 
-                // If the error is a single failed import then create a "fake" import
-                // resolution for it so that later resolve stages won't complain.
-                self.r.import_dummy_binding(import);
                 if prev_root_id.as_u32() != 0
                     && prev_root_id.as_u32() != import.root_id.as_u32()
                     && !errors.is_empty()
@@ -691,8 +699,6 @@ impl<'a, 'b> ImportResolver<'a, 'b> {
                     prev_root_id = import.root_id;
                 }
             } else if is_indeterminate {
-                // Consider erroneous imports used to avoid duplicate diagnostics.
-                self.r.used_imports.insert(import.id);
                 let path = import_path_to_string(
                     &import.module_path.iter().map(|seg| seg.ident).collect::<Vec<_>>(),
                     &import.kind,
@@ -824,25 +830,22 @@ impl<'a, 'b> ImportResolver<'a, 'b> {
                     Err(Undetermined) => indeterminate = true,
                     // Don't update the resolution, because it was never added.
                     Err(Determined) if target.name == kw::Underscore => {}
-                    Err(Determined) => {
+                    Ok(binding) if binding.is_importable() => {
+                        let imported_binding = this.import(binding, import);
+                        target_bindings[ns].set(Some(imported_binding));
+                        this.define(parent, target, ns, imported_binding);
+                    }
+                    source_binding @ (Ok(..) | Err(Determined)) => {
+                        if source_binding.is_ok() {
+                            let msg = format!("`{}` is not directly importable", target);
+                            struct_span_err!(this.session, import.span, E0253, "{}", &msg)
+                                .span_label(import.span, "cannot be imported directly")
+                                .emit();
+                        }
                         let key = this.new_key(target, ns);
                         this.update_resolution(parent, key, |_, resolution| {
                             resolution.single_imports.remove(&Interned::new_unchecked(import));
                         });
-                    }
-                    Ok(binding) if !binding.is_importable() => {
-                        let msg = format!("`{}` is not directly importable", target);
-                        struct_span_err!(this.session, import.span, E0253, "{}", &msg)
-                            .span_label(import.span, "cannot be imported directly")
-                            .emit();
-                        // Do not import this illegal binding. Import a dummy binding and pretend
-                        // everything is fine
-                        this.import_dummy_binding(import);
-                    }
-                    Ok(binding) => {
-                        let imported_binding = this.import(binding, import);
-                        target_bindings[ns].set(Some(imported_binding));
-                        this.define(parent, target, ns, imported_binding);
                     }
                 }
             }
@@ -876,10 +879,6 @@ impl<'a, 'b> ImportResolver<'a, 'b> {
             self.r.unusable_binding = orig_unusable_binding;
         }
         import.vis.set(orig_vis);
-        if let PathResult::Failed { .. } | PathResult::NonModule(..) = path_res {
-            // Consider erroneous imports used to avoid duplicate diagnostics.
-            self.r.used_imports.insert(import.id);
-        }
         let module = match path_res {
             PathResult::Module(module) => {
                 // Consistency checks, analogous to `finalize_macro_resolutions`.
@@ -1144,7 +1143,6 @@ impl<'a, 'b> ImportResolver<'a, 'b> {
                 })
             } else {
                 // `resolve_ident_in_module` reported a privacy error.
-                self.r.import_dummy_binding(import);
                 None
             };
         }

--- a/src/test/rustdoc-ui/intra-doc/unresolved-import-recovery.rs
+++ b/src/test/rustdoc-ui/intra-doc/unresolved-import-recovery.rs
@@ -1,0 +1,6 @@
+// Regression test for issue #95879.
+
+use unresolved_crate::module::Name; //~ ERROR failed to resolve
+
+/// [Name]
+pub struct S;

--- a/src/test/rustdoc-ui/intra-doc/unresolved-import-recovery.stderr
+++ b/src/test/rustdoc-ui/intra-doc/unresolved-import-recovery.stderr
@@ -1,0 +1,11 @@
+error[E0433]: failed to resolve: maybe a missing crate `unresolved_crate`?
+  --> $DIR/unresolved-import-recovery.rs:3:5
+   |
+LL | use unresolved_crate::module::Name;
+   |     ^^^^^^^^^^^^^^^^ maybe a missing crate `unresolved_crate`?
+
+error: Compilation failed, aborting rustdoc
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0433`.

--- a/src/test/ui/use/use-super-global-path.rs
+++ b/src/test/ui/use/use-super-global-path.rs
@@ -9,7 +9,7 @@ mod foo {
 
     pub fn g() {
         use ::super::main; //~ ERROR global paths cannot start with `super`
-        main(); //~ ERROR cannot find function `main` in this scope
+        main();
     }
 }
 

--- a/src/test/ui/use/use-super-global-path.stderr
+++ b/src/test/ui/use/use-super-global-path.stderr
@@ -16,18 +16,6 @@ error[E0433]: failed to resolve: global paths cannot start with `super`
 LL |         use ::super::main;
    |               ^^^^^ global paths cannot start with `super`
 
-error[E0425]: cannot find function `main` in this scope
-  --> $DIR/use-super-global-path.rs:12:9
-   |
-LL |         main();
-   |         ^^^^ not found in this scope
-   |
-help: consider importing this function
-   |
-LL |     use main;
-   |
+error: aborting due to 3 previous errors
 
-error: aborting due to 4 previous errors
-
-Some errors have detailed explanations: E0425, E0433.
-For more information about an error, try `rustc --explain E0425`.
+For more information about this error, try `rustc --explain E0433`.

